fix: update snap packaging

siggiskulason opened the pull request:
- Add support for defining provision watcher directory
- Add auto-scan script to snap
- Add support in auto-configure script/command for providing Consul token
- Update snap base to core 20
- Update documentation
- Update metadata and versioning
- Add support for "snap set" of device-specific settings
- Update install hook
- Update make section in snapcraft.yaml
- Other minor updates

Fix https://github.com/edgexfoundry/device-rfid-llrp-go/issues/55

See also: https://github.com/edgexfoundry/device-rfid-llrp-go/pull/52 for further information.

There is an issue with setting the subnet information using "snap set", which is unrelated to this PR - see edgexfoundry/app-functions-sdk-go#1043

Signed-off-by: Siggi Skulason <siggi.skulason@canonical.com>

<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/device-rfid-llrp-go/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-rfid-llrp-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?) N/A
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?) N/A
  <link to docs PR>

## Testing Instructions
<!-- How can the reviewers test your change? -->

Install snap and run device discovery with
```
sudo snap remove edgexfoundry
sudo snap remove edgex-device-rfid-llrp
sudo snap install edgexfoundry --edge

snapcraft try
sudo snap try prime

sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-rfid-llrp:edgex-secretstore-token
sudo snap start edgex-device-rfid-llrp.device-rfid-llrp 
sudo snap connect edgex-device-rfid-llrp:network-control 
sudo snap connect edgex-device-rfid-llrp:network-observe 

CONSUL_TOKEN=$(sudo cat /var/snap/edgexfoundry/current/secrets/consul-acl-token/bootstrap_token.json | jq ".SecretID" | tr -d '"') 
echo $CONSUL_TOKEN

sudo edgex-device-rfid-llrp.auto-configure $CONSUL_TOKEN

```

I have verified that this detects my RFID reader.


## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

siggiskulason commented:
@lenny-intel - Besides snap-specific changes, I made a couple of changes to `driver.go` and `auto-configure.sh` which you might want to review. Thanks.

farshidtz commented on snap/snapcraft.yaml:
```diff
@@ -85,43 +80,58 @@ parts:
+
+ 
+  # copy auto-configure.sh to /bin and make sure curl is installed
+  auto-configure:
+    plugin: dump
+    source: bin/
+    organize:
+      auto-configure.sh: bin/auto-configure.sh
```
This also copies `bin/test-attribution-txt.sh` and leaves it under $SNAP. Need to include only `bin/auto-configure.sh`, or if possible just source a single file `bin/auto-configure.sh`.

siggiskulason replied:
pushed an update with
```
  auto-configure:
    plugin: dump
    source: bin/
    filesets:
      auto-configure: [bin/auto-configure.sh]
    organize:
      auto-configure.sh: bin/auto-configure.sh
    stage-packages:
      - curl
 ```
which just copies this file

farshidtz reviewed (changes requested):
Thanks a lot @siggiskulason for the updates. I've added some inline comments. Apart from those, I suggest following updates:

1. Update `github.com/canonical/edgex-snap-hooks/v2` so that it includes bug fixes and better error logging (e.g. when setting wrong config option key).
2. The snap readme describes installation from 2.0 (Ireland) channel which doesn't exist. Please remove that part until we actually have a v2 channel and tag; see #63 
3. Reduce the content of "Using a content interface to set device configuration" section under snap readme to:
```
The `device-config` content interface allows another snap to seed this snap with configuration directories under `$SNAP_DATA/config/device-rfid-llrp`.

Note that the `device-config` content interface does NOT support seeding of the Secret Store Token because that file is expected at a different path.

Please refer to [edgex-config-provider](https://github.com/canonical/edgex-config-provider), for an example and further instructions.
```

lenny-intel commented on README.md:
```diff
@@ -35,10 +35,11 @@ make docker
 
 - Docker
-    - Use [compose-builder](https://github.com/edgexfoundry/edgex-compose/tree/jakarta/compose-builder)
+    - Use [compose-builder](https://github.com/edgexfoundry/edgex-compose/tree/main/compose-builder)
```
We are release this for Jakarta, so this needs to stay as is.

siggiskulason replied:
fixed.

lenny-intel commented on internal/driver/driver.go:
```diff
@@ -713,6 +712,13 @@ func getAddr(protocols protocolMap) (net.Addr, error) {
 
 func (d *Driver) addProvisionWatchers() error {
+
+	provisionWatcherFolder := d.config.AppCustom.ProvisionWatcherDir
```
This seems to be a missing feature in the SDK similar to `ProfilesDir` and `DevicesDir`. Please submit issue against the SDK (DS will need access to the setting) and add a TODO here to remove this once added to the SDK.

siggiskulason replied:
Comment added - and issue at https://github.com/edgexfoundry/device-sdk-go/issues/1100

lenny-intel replied:
Excellent. Thanks!

siggiskulason commented:
> * Update `github.com/canonical/edgex-snap-hooks/v2` so that it includes bug fixes and better error logging (e.g. when setting wrong config option key).

Updated to
```
require github.com/canonical/edgex-snap-hooks/v2 v2.1.3
```

> * The snap readme describes installation from 2.0 (Ireland) channel which doesn't exist. Please remove that part until we actually have a v2 channel and tag; see #63
> * Reduce the content of "Using a content interface to set device configuration" section under snap readme to:

Updated

siggiskulason commented on README.md:
```diff
@@ -10,16 +10,16 @@ A ROAccessReport can be used to examine data read from one or more RFID tags see
 
 ## Table of contents
-
+ 
 * [First Run](#first-run)
 * [Device Discovery](#device-discovery)
 * [Device Profiles, Custom LLRP Messages, and Service Limitations](#device-profiles-custom-llrp-messages-and-service-limitations)
 * [Connection Management](#connection-management)
 * [Example Scripts](#example-scripts)
 * [Testing](#testing)
-* [Snap Development and Testing](#snap-development-and-testing)
 * [Footnotes](#footnotes)
```
Fixed

siggiskulason commented:
@lenny-intel / @farshidtz - Thanks for the review. I have implemented all requested changes. Could you please re-review and approve if there are no other change requests? Thanks!

lenny-intel reviewed (approved):
LGTM, from my perspective. I'll let the other snap experts approve before merging.

farshidtz commented on snap/README.md:
````diff
@@ -69,40 +63,44 @@ ensures that as well as starting the service now, it will be automatically start
 
-To use, create a new snap with a directory containing the configuration files.
-Your `snapcraft.yaml` file then needs to define a slot with read access to the directory you are sharing.
+1. Using `snap set` to set your local subnet information. Example:
 
+```bash
+$ sudo snap set edgex-device-rfid-llrp env.app-custom.discovery-subnets="192.168.10.0/24"
+curl -X POST http://localhost:59989/api/v2/discovery
````
Please reference the appropriate Github issue here and mention that this will work only after the issue is resolved.

The issue is now in app-functions-sdk, which isn't even used by this service: https://github.com/edgexfoundry/app-functions-sdk-go/issues/1043

siggiskulason replied:
updated - please see latest commit.

farshidtz commented on snap/README.md:
```diff
@@ -69,40 +63,44 @@ ensures that as well as starting the service now, it will be automatically start
-where `$SNAP/config` is configuration directory your snap is providing to the device snap.
 
-Then connect the plug in the device snap to the slot in your snap, which will replace the configuration in the device snap. Do this with:
+2. Using the `auto-configure` command. This command finds all local network interfaces which are online and non-virtual and sets the value of `DiscoverySubnets` 
```
The numbers make it look like the orders matter. But in fact, this is an alternative to 1. Please improve.

farshidtz commented on snap/README.md:
```diff
@@ -145,28 +143,48 @@ For details on the mapping of configuration options to Config options, please re
-service.max-request-size        // Service.MaxRequestSize
-service.startup-msg             // Service.StartupMsg
-service.request-timeout         // Service.RequestTimeout
+service.boot-timeout                    // Service.BootTimeout
+service.health-check-interval           // Service.HealthCheckInterval
+service.host                            // Service.Host
+service.server-bind-addr                // Service.ServerBindAddr
+service.port                            // Service.Port
+service.protocol                        // Service.Protocol
```
`service.boot-timeout` and `service.protocol` aren't supported and are already removed from some other snaps readmes.

siggiskulason replied:
updated - please see latest commit.

siggiskulason commented on snap/snapcraft.yaml:
```diff
@@ -85,43 +79,60 @@ parts:
+    filesets:
+      auto-configure: [bin/auto-configure.sh]
+    organize:
+      auto-configure.sh: bin/auto-configure.sh
```
ah! thanks. updated - please see latest commit.

siggiskulason commented:
@farshidtz - I've pushed a commit with your requested changes. Please re-review.

farshidtz reviewed (approved):
Thanks. Looks good now. 

I don't have the actual device and reviewed only the packaging and runtime of the snap.

siggiskulason commented:
Thanks @lenny-intel / @farshidtz
